### PR TITLE
extract magic numbers to version catalog #16517

### LINKED ISSUE #16517: [Enhancement] Extract magic build numbers to Version Catalog
idle thought - is there a way to use the gradle version catalog to contain these magic numbers? (compileSdk / minSdk / targetSdk / compileOptions compatibility / kotlinOptions jvmTarget)

They are repeated a bunch

_Originally posted by @mikehardy in https://github.com/ankidroid/Anki-Android/pull/16498#discussion_r1621067466_

### PR DESCRIPTION
**Pull request by AbdelrahmanEsam**
extract magic numbers from gradle and share them between modules 

* Fixes #16517

### CODE REVIEW

File: AnkiDroid/build.gradle [hunk 1]
```diff
@@ -74,11 +74,11 @@ android {
         //
         // This ensures the correct ordering between the various types of releases (dev < alpha < beta < release) which is
         // needed for upgrades to be offered correctly.
-        versionCode=21900104
-        versionName="2.19alpha4"
-        minSdk 23 // also in testlib/build.gradle.kts
+        versionCode=libs.versions.versionCode.get().toInteger()
+        versionName=libs.versions.versionName.get().toString()
```
Comment (lukstbit) [inline AnkiDroid/build.gradle:78]: versionCode and versionName shouldn't be extracted.

**Review by lukstbit — CHANGES_REQUESTED** [review]
Thanks, left one minor issue.
Also please link the issue that this PR will be closing.

File: AnkiDroid/build.gradle (hunk 1 above)
Comment (lukstbit) [reply AnkiDroid/build.gradle:78]: Those are used just in the main module to build the app and they will not be used in other modules. Also, having the actual values for the app versions there, is better for understanding the documentation that is just above the changed lines(that describes how the versions numbers are constructed).

The other values that you extracted are used in multiple modules(and will be used in modules that we plan to add).

Comment (AbdelrahmanEsam) [reply AnkiDroid/build.gradle:78]: I see that they are only in the main module and they will never be at another ... but I see that it's good for centerlizing versions over all .... you want to alter version ? just go for version catalog.

Comment (AbdelrahmanEsam) [reply AnkiDroid/build.gradle:78]: @david-allison  what do you think ?

Comment (david-allison) [reply AnkiDroid/build.gradle:78]: Keep the version codes where they are

File: AnkiDroid/build.gradle [hunk 2]
```diff
@@ -35,7 +35,7 @@ static def gitCommitHash() {
 android {
     namespace "com.ichi2.anki"
 
-    compileSdk 34 // change api compileSdk at the same time
+    compileSdk libs.versions.compileSdk.get().toInteger() // change api compileSdk at the same time
```
Comment (david-allison) [inline AnkiDroid/build.gradle:38]:
```suggestion
    compileSdk libs.versions.compileSdk.get().toInteger()
```

File: AnkiDroid/build.gradle [hunk 3]
```diff
@@ -76,9 +76,9 @@ android {
         // needed for upgrades to be offered correctly.
         versionCode=21900104
         versionName="2.19alpha4"
-        minSdk 23 // also in testlib/build.gradle.kts
+        minSdk libs.versions.minSdk.get().toInteger() // also in testlib/build.gradle.kts
```
Comment (david-allison) [inline AnkiDroid/build.gradle:79]:
```suggestion
        minSdk libs.versions.minSdk.get().toInteger()
```

File: AnkiDroid/build.gradle [hunk 4]
```diff
@@ -76,9 +76,9 @@ android {
...
         // After #13695: change .tests_emulator.yml
-        targetSdk 33 // also in [api|testlib]/build.gradle.kts and ../robolectricDownloader.gradle
+        targetSdk libs.versions.targetSdk.get().toInteger() // also in [api|testlib]/build.gradle.kts and ../robolectricDownloader.gradle
```
Comment (david-allison) [inline AnkiDroid/build.gradle:81]: Can the toml value also be used in `robolectricDownloader`?

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/build.gradle (hunk 4 above)
Comment (AbdelrahmanEsam) [reply AnkiDroid/build.gradle:81]: yes but for what ?

Comment (david-allison) [reply AnkiDroid/build.gradle]: You're right - forgot the implementation of the script

File: AnkiDroid/build.gradle [hunk 5]
```diff
@@ -76,9 +76,9 @@ android {
         // needed for upgrades to be offered correctly.
         versionCode=21900104
         versionName="2.19alpha4"
-        minSdk 23 // also in testlib/build.gradle.kts
+        minSdk libs.versions.minSdk.get().toInteger()
         // After #13695: change .tests_emulator.yml
-        targetSdk 33 // also in [api|testlib]/build.gradle.kts and ../robolectricDownloader.gradle
```
Comment (david-allison) [inline AnkiDroid/build.gradle:81]: move the `robolectricDownloader.gradle` comment to the TOML

**Review by david-allison — CHANGES_REQUESTED** [review]
one chage, as the result of a discussion - rest looks good, thanks!

Treat as an approve after the comment is added

File: AnkiDroid/build.gradle [hunk 6]
```diff
@@ -76,9 +76,9 @@ android {
         // needed for upgrades to be offered correctly.
         versionCode=21900104
         versionName="2.19alpha4"
-        minSdk 23 // also in testlib/build.gradle.kts
+        minSdk libs.versions.minSdk.get().toInteger()
         // After #13695: change .tests_emulator.yml
```
Comment (david-allison) [inline AnkiDroid/build.gradle:80]: maybe also move this to `minSdk`

**Review by david-allison — APPROVED** [review]

File: gradle/libs.versions.toml [hunk 7]
```diff
@@ -1,4 +1,7 @@
 [versions]
+compileSdk = "34"
+minSdk = "23"
+targetSdk = "33"  # also in [api|testlib]/build.gradle.kts and ../robolectricDownloader.gradle
```
Comment (david-allison) [inline gradle/libs.versions.toml:4]:
```suggestion
targetSdk = "33"  # also in ../robolectricDownloader.gradle
```

**Review by david-allison — APPROVED** [review]
CHeers

**Review by lukstbit — APPROVED** [review]
Thank you for making the changes!

**Pull request merged**